File: go/core/guild-wars2_sdk.go
package core

import (
	"fmt"

	vs "github.com/voxgig-sdk/guild-wars2-sdk/go/utility/struct"
)

type GuildWars2SDK struct {
	Mode     string
	options  map[string]any
	utility  *Utility
	Features []Feature
	rootctx  *Context
}

func NewGuildWars2SDK(options map[string]any) *GuildWars2SDK {
	sdk := &GuildWars2SDK{
		Mode:     "live",
		Features: []Feature{},
	}

	sdk.utility = NewUtility()

	config := MakeConfig()

	sdk.rootctx = sdk.utility.MakeContext(map[string]any{
		"client":  sdk,
		"utility": sdk.utility,
		"config":  config,
		"options": options,
		"shared":  map[string]any{},
	}, nil)

	sdk.options = sdk.utility.MakeOptions(sdk.rootctx)

	if vs.GetPath([]any{"feature", "test", "active"}, sdk.options) == true {
		sdk.Mode = "test"
	}

	sdk.rootctx.Options = sdk.options

	// Add features from config.
	featureOpts := ToMapAny(vs.GetProp(sdk.options, "feature"))
	if featureOpts != nil {
		for _, item := range vs.Items(featureOpts) {
			fname, _ := item[0].(string)
			fopts := ToMapAny(item[1])
			if fopts != nil {
				if active, ok := fopts["active"]; ok {
					if ab, ok := active.(bool); ok && ab {
						sdk.utility.FeatureAdd(sdk.rootctx, makeFeature(fname))
					}
				}
			}
		}
	}

	// Add extension features.
	if extend := vs.GetProp(sdk.options, "extend"); extend != nil {
		if extList, ok := extend.([]any); ok {
			for _, f := range extList {
				if feat, ok := f.(Feature); ok {
					sdk.utility.FeatureAdd(sdk.rootctx, feat)
				}
			}
		}
	}

	// Initialize features.
	for _, f := range sdk.Features {
		sdk.utility.FeatureInit(sdk.rootctx, f)
	}

	sdk.utility.FeatureHook(sdk.rootctx, "PostConstruct")

	return sdk
}

func (sdk *GuildWars2SDK) OptionsMap() map[string]any {
	out := vs.Clone(sdk.options)
	if om, ok := out.(map[string]any); ok {
		return om
	}
	return map[string]any{}
}

func (sdk *GuildWars2SDK) GetUtility() *Utility {
	return CopyUtility(sdk.utility)
}

func (sdk *GuildWars2SDK) GetRootCtx() *Context {
	return sdk.rootctx
}

func (sdk *GuildWars2SDK) Prepare(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "prepare",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	options := sdk.options

	path, _ := vs.GetProp(fetchargs, "path").(string)
	method, _ := vs.GetProp(fetchargs, "method").(string)
	if method == "" {
		method = "GET"
	}

	params := ToMapAny(vs.GetProp(fetchargs, "params"))
	if params == nil {
		params = map[string]any{}
	}
	query := ToMapAny(vs.GetProp(fetchargs, "query"))
	if query == nil {
		query = map[string]any{}
	}

	headers := utility.PrepareHeaders(ctx)

	base, _ := vs.GetProp(options, "base").(string)
	prefix, _ := vs.GetProp(options, "prefix").(string)
	suffix, _ := vs.GetProp(options, "suffix").(string)

	ctx.Spec = NewSpec(map[string]any{
		"base":    base,
		"prefix":  prefix,
		"suffix":  suffix,
		"path":    path,
		"method":  method,
		"params":  params,
		"query":   query,
		"headers": headers,
		"body":    vs.GetProp(fetchargs, "body"),
		"step":    "start",
	})

	// Merge user-provided headers.
	if uh := vs.GetProp(fetchargs, "headers"); uh != nil {
		if uhm, ok := uh.(map[string]any); ok {
			for k, v := range uhm {
				ctx.Spec.Headers[k] = v
			}
		}
	}

	_, err := utility.PrepareAuth(ctx)
	if err != nil {
		return nil, err
	}

	return utility.MakeFetchDef(ctx)
}

func (sdk *GuildWars2SDK) Direct(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	fetchdef, err := sdk.Prepare(fetchargs)
	if err != nil {
		return map[string]any{"ok": false, "err": err}, nil
	}

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "direct",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	url, _ := fetchdef["url"].(string)
	fetched, fetchErr := utility.Fetcher(ctx, url, fetchdef)

	if fetchErr != nil {
		return map[string]any{"ok": false, "err": fetchErr}, nil
	}

	if fetched == nil {
		return map[string]any{
			"ok":  false,
			"err": ctx.MakeError("direct_no_response", "response: undefined"),
		}, nil
	}

	if fm, ok := fetched.(map[string]any); ok {
		status := ToInt(vs.GetProp(fm, "status"))
		headers := vs.GetProp(fm, "headers")

		// No-body responses (204, 304) and explicit zero content-length
		// must skip JSON parsing — calling json() on an empty body errors.
		var contentLength string
		if hm, ok := headers.(map[string]any); ok {
			if cl, ok := hm["content-length"]; ok {
				contentLength = fmt.Sprintf("%v", cl)
			}
		}
		noBody := status == 204 || status == 304 || contentLength == "0"

		var jsonData any
		if !noBody {
			if jf := vs.GetProp(fm, "json"); jf != nil {
				if f, ok := jf.(func() any); ok {
					// f() returns nil on parse error in our fetcher.
					jsonData = f()
				}
			}
		}

		return map[string]any{
			"ok":      status >= 200 && status < 300,
			"status":  status,
			"headers": headers,
			"data":    jsonData,
		}, nil
	}

	return map[string]any{"ok": false, "err": ctx.MakeError("direct_invalid", "invalid response type")}, nil
}


// Achievement returns a Achievement entity bound to this client.
// Idiomatic usage: client.Achievement(nil).List(nil, nil) or
// client.Achievement(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) Achievement(data map[string]any) GuildWars2Entity {
	return NewAchievementEntityFunc(sdk, data)
}


// Authenticated returns a Authenticated entity bound to this client.
// Idiomatic usage: client.Authenticated(nil).List(nil, nil) or
// client.Authenticated(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) Authenticated(data map[string]any) GuildWars2Entity {
	return NewAuthenticatedEntityFunc(sdk, data)
}


// DailyReward returns a DailyReward entity bound to this client.
// Idiomatic usage: client.DailyReward(nil).List(nil, nil) or
// client.DailyReward(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) DailyReward(data map[string]any) GuildWars2Entity {
	return NewDailyRewardEntityFunc(sdk, data)
}


// GameMechanic returns a GameMechanic entity bound to this client.
// Idiomatic usage: client.GameMechanic(nil).List(nil, nil) or
// client.GameMechanic(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) GameMechanic(data map[string]any) GuildWars2Entity {
	return NewGameMechanicEntityFunc(sdk, data)
}


// Guild returns a Guild entity bound to this client.
// Idiomatic usage: client.Guild(nil).List(nil, nil) or
// client.Guild(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) Guild(data map[string]any) GuildWars2Entity {
	return NewGuildEntityFunc(sdk, data)
}


// GuildAuthenticated returns a GuildAuthenticated entity bound to this client.
// Idiomatic usage: client.GuildAuthenticated(nil).List(nil, nil) or
// client.GuildAuthenticated(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) GuildAuthenticated(data map[string]any) GuildWars2Entity {
	return NewGuildAuthenticatedEntityFunc(sdk, data)
}


// HomeInstance returns a HomeInstance entity bound to this client.
// Idiomatic usage: client.HomeInstance(nil).List(nil, nil) or
// client.HomeInstance(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) HomeInstance(data map[string]any) GuildWars2Entity {
	return NewHomeInstanceEntityFunc(sdk, data)
}


// Item returns a Item entity bound to this client.
// Idiomatic usage: client.Item(nil).List(nil, nil) or
// client.Item(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) Item(data map[string]any) GuildWars2Entity {
	return NewItemEntityFunc(sdk, data)
}


// Map returns a Map entity bound to this client.
// Idiomatic usage: client.Map(nil).List(nil, nil) or
// client.Map(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) Map(data map[string]any) GuildWars2Entity {
	return NewMapEntityFunc(sdk, data)
}


// MapInformation returns a MapInformation entity bound to this client.
// Idiomatic usage: client.MapInformation(nil).List(nil, nil) or
// client.MapInformation(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) MapInformation(data map[string]any) GuildWars2Entity {
	return NewMapInformationEntityFunc(sdk, data)
}


// Miscellaneous returns a Miscellaneous entity bound to this client.
// Idiomatic usage: client.Miscellaneous(nil).List(nil, nil) or
// client.Miscellaneous(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) Miscellaneous(data map[string]any) GuildWars2Entity {
	return NewMiscellaneousEntityFunc(sdk, data)
}


// Story returns a Story entity bound to this client.
// Idiomatic usage: client.Story(nil).List(nil, nil) or
// client.Story(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) Story(data map[string]any) GuildWars2Entity {
	return NewStoryEntityFunc(sdk, data)
}


// StructuredPvP returns a StructuredPvP entity bound to this client.
// Idiomatic usage: client.StructuredPvP(nil).List(nil, nil) or
// client.StructuredPvP(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) StructuredPvP(data map[string]any) GuildWars2Entity {
	return NewStructuredPvPEntityFunc(sdk, data)
}


// TradingPost returns a TradingPost entity bound to this client.
// Idiomatic usage: client.TradingPost(nil).List(nil, nil) or
// client.TradingPost(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) TradingPost(data map[string]any) GuildWars2Entity {
	return NewTradingPostEntityFunc(sdk, data)
}


// WorldVsWorld returns a WorldVsWorld entity bound to this client.
// Idiomatic usage: client.WorldVsWorld(nil).List(nil, nil) or
// client.WorldVsWorld(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *GuildWars2SDK) WorldVsWorld(data map[string]any) GuildWars2Entity {
	return NewWorldVsWorldEntityFunc(sdk, data)
}



func TestSDK(testopts map[string]any, sdkopts map[string]any) *GuildWars2SDK {
	if sdkopts == nil {
		sdkopts = map[string]any{}
	}
	sdkopts = vs.Clone(sdkopts).(map[string]any)

	if testopts == nil {
		testopts = map[string]any{}
	}
	testopts = vs.Clone(testopts).(map[string]any)
	testopts["active"] = true

	vs.SetPath(sdkopts, []any{"feature", "test"}, testopts)

	sdk := NewGuildWars2SDK(sdkopts)
	sdk.Mode = "test"

	return sdk
}

File: ts/src/GuildWars2SDK.ts
// GuildWars2 Ts SDK

import { AchievementEntity } from './entity/AchievementEntity'
import { AuthenticatedEntity } from './entity/AuthenticatedEntity'
import { DailyRewardEntity } from './entity/DailyRewardEntity'
import { GameMechanicEntity } from './entity/GameMechanicEntity'
import { GuildEntity } from './entity/GuildEntity'
import { GuildAuthenticatedEntity } from './entity/GuildAuthenticatedEntity'
import { HomeInstanceEntity } from './entity/HomeInstanceEntity'
import { ItemEntity } from './entity/ItemEntity'
import { MapEntity } from './entity/MapEntity'
import { MapInformationEntity } from './entity/MapInformationEntity'
import { MiscellaneousEntity } from './entity/MiscellaneousEntity'
import { StoryEntity } from './entity/StoryEntity'
import { StructuredPvPEntity } from './entity/StructuredPvPEntity'
import { TradingPostEntity } from './entity/TradingPostEntity'
import { WorldVsWorldEntity } from './entity/WorldVsWorldEntity'

export type * from './GuildWars2Types'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { GuildWars2EntityBase } from './GuildWars2EntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class GuildWars2SDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath
    const items = struct.items

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    items(this._options.feature, (fitem: [string, any]) => {
      const fname = fitem[0]
      const fopts = fitem[1]
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    })

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  _achievement?: AchievementEntity

  // Idiomatic facade: `client.achievement.list()` / `client.achievement.load({ id })`.
  get achievement(): AchievementEntity {
    return (this._achievement ??= new AchievementEntity(this, undefined))
  }

  /** @deprecated Use `client.achievement` instead. */
  Achievement(data?: any) {
    const self = this
    return new AchievementEntity(self,data)
  }


  _authenticated?: AuthenticatedEntity

  // Idiomatic facade: `client.authenticated.list()` / `client.authenticated.load({ id })`.
  get authenticated(): AuthenticatedEntity {
    return (this._authenticated ??= new AuthenticatedEntity(this, undefined))
  }

  /** @deprecated Use `client.authenticated` instead. */
  Authenticated(data?: any) {
    const self = this
    return new AuthenticatedEntity(self,data)
  }


  _daily_reward?: DailyRewardEntity

  // Idiomatic facade: `client.daily_reward.list()` / `client.daily_reward.load({ id })`.
  get daily_reward(): DailyRewardEntity {
    return (this._daily_reward ??= new DailyRewardEntity(this, undefined))
  }

  /** @deprecated Use `client.daily_reward` instead. */
  DailyReward(data?: any) {
    const self = this
    return new DailyRewardEntity(self,data)
  }


  _game_mechanic?: GameMechanicEntity

  // Idiomatic facade: `client.game_mechanic.list()` / `client.game_mechanic.load({ id })`.
  get game_mechanic(): GameMechanicEntity {
    return (this._game_mechanic ??= new GameMechanicEntity(this, undefined))
  }

  /** @deprecated Use `client.game_mechanic` instead. */
  GameMechanic(data?: any) {
    const self = this
    return new GameMechanicEntity(self,data)
  }


  _guild?: GuildEntity

  // Idiomatic facade: `client.guild.list()` / `client.guild.load({ id })`.
  get guild(): GuildEntity {
    return (this._guild ??= new GuildEntity(this, undefined))
  }

  /** @deprecated Use `client.guild` instead. */
  Guild(data?: any) {
    const self = this
    return new GuildEntity(self,data)
  }


  _guild_authenticated?: GuildAuthenticatedEntity

  // Idiomatic facade: `client.guild_authenticated.list()` / `client.guild_authenticated.load({ id })`.
  get guild_authenticated(): GuildAuthenticatedEntity {
    return (this._guild_authenticated ??= new GuildAuthenticatedEntity(this, undefined))
  }

  /** @deprecated Use `client.guild_authenticated` instead. */
  GuildAuthenticated(data?: any) {
    const self = this
    return new GuildAuthenticatedEntity(self,data)
  }


  _home_instance?: HomeInstanceEntity

  // Idiomatic facade: `client.home_instance.list()` / `client.home_instance.load({ id })`.
  get home_instance(): HomeInstanceEntity {
    return (this._home_instance ??= new HomeInstanceEntity(this, undefined))
  }

  /** @deprecated Use `client.home_instance` instead. */
  HomeInstance(data?: any) {
    const self = this
    return new HomeInstanceEntity(self,data)
  }


  _item?: ItemEntity

  // Idiomatic facade: `client.item.list()` / `client.item.load({ id })`.
  get item(): ItemEntity {
    return (this._item ??= new ItemEntity(this, undefined))
  }

  /** @deprecated Use `client.item` instead. */
  Item(data?: any) {
    const self = this
    return new ItemEntity(self,data)
  }


  _map?: MapEntity

  // Idiomatic facade: `client.map.list()` / `client.map.load({ id })`.
  get map(): MapEntity {
    return (this._map ??= new MapEntity(this, undefined))
  }

  /** @deprecated Use `client.map` instead. */
  Map(data?: any) {
    const self = this
    return new MapEntity(self,data)
  }


  _map_information?: MapInformationEntity

  // Idiomatic facade: `client.map_information.list()` / `client.map_information.load({ id })`.
  get map_information(): MapInformationEntity {
    return (this._map_information ??= new MapInformationEntity(this, undefined))
  }

  /** @deprecated Use `client.map_information` instead. */
  MapInformation(data?: any) {
    const self = this
    return new MapInformationEntity(self,data)
  }


  _miscellaneous?: MiscellaneousEntity

  // Idiomatic facade: `client.miscellaneous.list()` / `client.miscellaneous.load({ id })`.
  get miscellaneous(): MiscellaneousEntity {
    return (this._miscellaneous ??= new MiscellaneousEntity(this, undefined))
  }

  /** @deprecated Use `client.miscellaneous` instead. */
  Miscellaneous(data?: any) {
    const self = this
    return new MiscellaneousEntity(self,data)
  }


  _story?: StoryEntity

  // Idiomatic facade: `client.story.list()` / `client.story.load({ id })`.
  get story(): StoryEntity {
    return (this._story ??= new StoryEntity(this, undefined))
  }

  /** @deprecated Use `client.story` instead. */
  Story(data?: any) {
    const self = this
    return new StoryEntity(self,data)
  }


  _structured_pv_p?: StructuredPvPEntity

  // Idiomatic facade: `client.structured_pv_p.list()` / `client.structured_pv_p.load({ id })`.
  get structured_pv_p(): StructuredPvPEntity {
    return (this._structured_pv_p ??= new StructuredPvPEntity(this, undefined))
  }

  /** @deprecated Use `client.structured_pv_p` instead. */
  StructuredPvP(data?: any) {
    const self = this
    return new StructuredPvPEntity(self,data)
  }


  _trading_post?: TradingPostEntity

  // Idiomatic facade: `client.trading_post.list()` / `client.trading_post.load({ id })`.
  get trading_post(): TradingPostEntity {
    return (this._trading_post ??= new TradingPostEntity(this, undefined))
  }

  /** @deprecated Use `client.trading_post` instead. */
  TradingPost(data?: any) {
    const self = this
    return new TradingPostEntity(self,data)
  }


  _world_vs_world?: WorldVsWorldEntity

  // Idiomatic facade: `client.world_vs_world.list()` / `client.world_vs_world.load({ id })`.
  get world_vs_world(): WorldVsWorldEntity {
    return (this._world_vs_world ??= new WorldVsWorldEntity(this, undefined))
  }

  /** @deprecated Use `client.world_vs_world` instead. */
  WorldVsWorld(data?: any) {
    const self = this
    return new WorldVsWorldEntity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new GuildWars2SDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return GuildWars2SDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'GuildWars2' }
  }

  toString() {
    return 'GuildWars2 ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = GuildWars2SDK


export {
  stdutil,

  BaseFeature,
  GuildWars2EntityBase,

  GuildWars2SDK,
  SDK,
}


